upf/upf_cli.c

```diff
@@ -635,6 +634,85 @@ VLIB_CLI_COMMAND (upf_show_spec_release_command, static) =
 };
 /* *INDENT-ON* */
 
+static clib_error_t *
+upf_node_id_command_fn (vlib_main_t * vm,
+			unformat_input_t * main_input,
+			vlib_cli_command_t * cmd)
+{
+  upf_main_t *gtm = &upf_main;
+  unformat_input_t _line_input, *line_input = &_line_input;
+  clib_error_t *error = NULL;
+  ip46_address_t ip46;
+  u8 *fqdn = 0;
+  pfcp_node_id_t *node_id = &gtm->node_id;
+
+  if (!unformat_user (main_input, unformat_line_input, line_input))
+    return 0;
+
+  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
+    {
+      if (unformat (line_input, "fqdn %_%v%_", &fqdn))
+	{
+	  free_node_id (node_id);
+	  node_id->type = NID_FQDN;
+	  node_id->fqdn = upf_name_to_labels (fqdn);
+	  vec_free (fqdn);
+	}
+      else if (unformat (line_input, "ip4 %U",
+			 unformat_ip46_address, &ip46, IP46_TYPE_ANY))
+	{
+	  free_node_id (node_id);
+	  node_id->type = NID_IPv4;
+	  node_id->ip = ip46;
+	}
+      else if (unformat (line_input, "ip6 %U",
+			 unformat_ip46_address, &ip46, IP46_TYPE_ANY))
+	{
+	  free_node_id (node_id);
+	  node_id->type = NID_IPv6;
+	  node_id->ip = ip46;
+	}
+      else
+	{
+	  error = unformat_parse_error (line_input);
+	  return error;
+	}
+    }
+
+  return NULL;
+
+}
```

Comment:
what about this instead?

```suggestion
static clib_error_t *
upf_node_id_command_fn (vlib_main_t * vm,
			unformat_input_t * main_input,
			vlib_cli_command_t * cmd)
{
  upf_main_t *gtm = &upf_main;
  unformat_input_t _line_input, *line_input = &_line_input;
  clib_error_t *error = NULL;
  pfcp_node_id_t node_id;
  u8 *fqdn = 0;

  if (!unformat_user (main_input, unformat_line_input, line_input))
    return 0;

  while (unformat_check_input (line_input) != UNFORMAT_END_OF_INPUT)
    {
      if (unformat (line_input, "fqdn %_%v%_", &fqdn))
	{
	  node_id.type = NID_FQDN;
	  node_id.fqdn = upf_name_to_labels (fqdn);
	  vec_free (fqdn);
	}
      else if (unformat (line_input, "ip4 %U",
			 unformat_ip46_address, &node_id->ip, IP46_TYPE_IP4))
	{
	  node_id.type = NID_IPv4;
	}
      else if (unformat (line_input, "ip6 %U",
			 unformat_ip46_address, &node_id->ip, IP46_TYPE_IP6))
	{
	  node_id.type = NID_IPv6;
	}
      else
	{
	  error = unformat_parse_error (line_input);
	  return error;
	}
    }

  free_node_id (&gtm->node_id);
  gtm->node_id = node_id;
  return NULL;
}
```

Reply:
Thanks for hit
refactored